### DISCUSSION TITLE
Seeking Guidance: Issues with Raspberry Pi & SAS Device Communication via USB to Serial Adapter

### DISCUSSION THREAD

**Discussion by jkewtts92** [zacharytomlinson/saspy#36] [Q&A] [answered]
First off, I want to express my deep appreciation for the work that has been shared here. It's been incredibly helpful.

I'm currently facing a persistent challenge in my project. Despite experimenting with various strategies. I'm encountering a recurring problem that I haven't been able to resolve. The issue centers around communication between a Raspberry Pi and a SAS device using a USB to serial adapter.

Here are the specifics:

Methods Attempted: I've implemented and tested multiple approaches discussed here, focusing particularly on the start and _send_command methods.
Current Outcome: Although I'm able to establish a connection to the SAS device, the self.connection.read(size) line consistently returns an empty response, which is puzzling. This ripples through the rest of the methods. and the start method returns 0c.
Given this situation, I would greatly appreciate any insights, advice, or suggestions on how to proceed. I'm particularly interested in:

Any potential reasons for receiving an empty response despite a successful connection.

Adjustments or alternative approaches that could potentially solve this issue.

Additional diagnostics or tests that could help pinpoint the underlying problem.
Thank you in advance for your time and assistance.

**Comment by well-it-wasnt-me** [comment]
Hi there ! Thank you so much for the kind words ! Really appreciated. 

Could you provide us first few more details ? Like: 

- vendor and model of the machine
- which commands are you trying to send 
- which version of saspy are you using 
- if possible a snippet of the code. 


By the way, and I'm relying on my memory...so not a very trusted source, there might be some commands that do not have a return but are checked in other ways if are successful

**Reply by jkewtts92** [reply]
ironically I figured out the issue right after I posted. 

Again can't thank you enough. 

just getting started with this protocol so may be reaching out in the future.

**Reply by well-it-wasnt-me** [reply]
No problem. We're always here. 

Do you mind sharing the solution/origin of the problem ? This would help other people that might encounter/do same mistake

**Reply by jkewtts92** [reply, marked as answer]
Port Configuration: Modified the Port setting within the config.yml file to align with the correct communication port requirements.

Configuration Path: Updated the config_handler file to point accurately to the location of the config.yml file, ensuring the application can correctly access and parse the configuration settings.

Code Refinements:
Incorporated the sas.start() method at the beginning of my run file, as highlighted by the example.py file. This step was crucial for initializing the system correctly before invoking other methods, addressing startup sequence issues.

Adjustment in sas.py: Identified and rectified an error in the sas.start() method. The specific line self.address = int(binascii.hexlify(response)) was failing due to improper base specification for the conversion. Adding ,16 to specify hexadecimal conversion (self.address = int(binascii.hexlify(response), 16)) resolved this issue, allowing the application to process the response correctly.

**Reply by well-it-wasnt-me** [reply]
Thank you ! We'll update accordingly!
